ogr/ogrsf_frmts/miramon/mm_gdal_functions.c

```diff
@@ -0,0 +1,2801 @@
+{
+    MM_N_HEIGHT_TYPE i;
+    MM_N_VERTICES_TYPE i_vrt;
+    double *pcoord_z;
+    MM_N_HEIGHT_TYPE n_alcada, n_h_total;
+    int tipus;
+    double *alcada = nullptr, *palcada, *palcada_i;
+#define MM_N_ALCADA_LOCAL 50  // Nr of local heights
+    double local_CinquantaAlcades[MM_N_ALCADA_LOCAL];
+
+    for (i_vrt = 0; i_vrt < n_vrt; i_vrt++)
+        coord_z[i_vrt] = MM_NODATA_COORD_Z;
+
+    tipus = MM_ARC_HEIGHT_TYPE(pZDescription->nZCount);
+    n_alcada = MM_ARC_N_HEIGHTS(pZDescription->nZCount);
+    if (n_vrt == 0 || n_alcada == 0)
+        return 0;
+
+    if (tipus == MM_ARC_HEIGHT_FOR_EACH_VERTEX)
+        n_h_total = (MM_N_HEIGHT_TYPE)n_vrt * n_alcada;
+    else
+        n_h_total = n_alcada;
+
+    if (n_h_total <= MM_N_ALCADA_LOCAL)
+        palcada = local_CinquantaAlcades;
+    else
+    {
+        if (MMCheckSize_t(n_vrt * sizeof(double) * n_alcada, 1))
+            return 1;
+        if (nullptr == (palcada = alcada = calloc_function(
+                            (size_t)n_vrt * sizeof(double) * n_alcada)))
+            return 1;
+    }
+
+    if (fseek_function(pF, pZDescription->nOffsetZ, SEEK_SET))
+    {
+        if (alcada)
+            free_function(alcada);
+        return 1;
+    }
+    if (n_h_total != (MM_N_HEIGHT_TYPE)fread_function(palcada, sizeof(double),
+                                                      n_h_total, pF))
+    {
+        if (alcada)
+            free_function(alcada);
+        return 1;
+    }
+
+    if (tipus == MM_ARC_HEIGHT_FOR_EACH_VERTEX)
+    {
+        palcada_i = palcada;
+        for (i = 0; i < n_alcada; i++)
+        {
+            for (i_vrt = 0, pcoord_z = coord_z; i_vrt < n_vrt;
+                 i_vrt++, pcoord_z++, palcada_i++)
+                MM_AdoptHeight(pcoord_z, palcada_i, flag);
+        }
+    }
+    else
+    {
+        palcada_i = palcada;
+        pcoord_z = coord_z;
+        for (i = 0; i < n_alcada; i++, palcada_i++)
+            MM_AdoptHeight(pcoord_z, palcada_i, flag);
+
+        if (*pcoord_z != MM_NODATA_COORD_Z)
+        {
+            /*Copio el mateix valor a totes les alcades.*/
+            for (i_vrt = 1, pcoord_z++; i_vrt < (size_t)n_vrt;
+                 i_vrt++, pcoord_z++)
+                *pcoord_z = *coord_z;
+        }
+    }
+    if (alcada)
+        free_function(alcada);
+    return 0;
+}  // End of MM_GetArcHeights()
+
+static char *MM_l_RemoveWhitespacesFromEndOfString(char *punter,
+                                                   size_t l_cadena)
+{
+    int longitud_cadena = (int)l_cadena;
+    if (longitud_cadena-- == 0)
+        return punter;
+
+    if (punter[longitud_cadena] != ' ' && punter[longitud_cadena] != '\t')
+        return punter;
+    longitud_cadena--;
+
+    while (longitud_cadena > -1)
+    {
+        if (punter[longitud_cadena] != ' ' && punter[longitud_cadena] != '\t')
+        {
+            break;
+        }
+        longitud_cadena--;
+    }
+
+    punter[++longitud_cadena] = '\0';
```

Comment:
I believe this can be rewritten in a more compact way and without any cast:

```suggestion
    size_t longitud_cadena = l_cadena;
    while (longitud_cadena > 0)
    {
        longitud_cadena--;
        if (punter[longitud_cadena] != ' ' && punter[longitud_cadena] != '\t')
        {
            break;
        }
        punter[longitud_cadena] = '\0';
    }
```

Reply:
Done